src/main/scala/it/pagopa/interop/agreementprocess/service/CatalogManagementService.scala

```diff
@@ -15,79 +21,55 @@ trait CatalogManagementService {
-    val allowedStatus: List[EServiceDescriptorState] = List(EServiceDescriptorState.PUBLISHED)
+  def validateActivationOnDescriptor(eservice: EService, descriptorId: UUID): Future[Unit] = {
+    val allowedStatus: List[EServiceDescriptorState] =
+      List(EServiceDescriptorState.PUBLISHED, EServiceDescriptorState.DEPRECATED, EServiceDescriptorState.SUSPENDED)
     validateEServiceDescriptorStatus(eservice, descriptorId, allowedStatus)
   }
-
-  def validateOperationOnDescriptor(eservice: EService, descriptorId: UUID): Future[EService] = {
+  def validateCreationOnDescriptor(eservice: EService, descriptorId: UUID): Future[Unit]   = {
     val allowedStatus: List[EServiceDescriptorState] =
-      List(EServiceDescriptorState.PUBLISHED, EServiceDescriptorState.DEPRECATED)
+      List(EServiceDescriptorState.PUBLISHED, EServiceDescriptorState.SUSPENDED)
+    validateEServiceDescriptorStatus(eservice, descriptorId, allowedStatus)
+  }
+  def validateSubmitOnDescriptor(eservice: EService, descriptorId: UUID): Future[Unit]     = {
+    val allowedStatus: List[EServiceDescriptorState] =
+      List(EServiceDescriptorState.PUBLISHED, EServiceDescriptorState.SUSPENDED)
     validateEServiceDescriptorStatus(eservice, descriptorId, allowedStatus)
   }
 
   def validateEServiceDescriptorStatus(
-    eservice: EService,
+    eService: EService,
     descriptorId: UUID,
-    allowedStatus: List[EServiceDescriptorState]
-  ): Future[EService] = {
-    val descriptorStatus = eservice.descriptors.find(_.id == descriptorId).map(_.state)
-
-    Future.fromTry(
-      Either
-        .cond(
-          descriptorStatus.exists(status => allowedStatus.contains(status)),
-          eservice,
-          new RuntimeException(
-            s"Descriptor ${descriptorId.toString} of Eservice ${eservice.id} has not status in ${allowedStatus
-                .mkString("[", ",", "]")}"
-          )
-        )
-        .toTry
-    )
-  }
-
-  def getDescriptorAudience(eservice: EService, descriptorId: UUID): Future[Seq[String]] = {
-    val audience = eservice.descriptors.find(_.id == descriptorId).map(_.audience)
-
-    audience match {
-      case Some(aud) => Future.successful[Seq[String]](aud)
-      case None      =>
-        Future.failed[Seq[String]](DescriptorNotFound(eservice.id.toString, descriptorId.toString))
-    }
-  }
+    allowedStates: List[EServiceDescriptorState]
+  ): Future[Unit] = {
+    val descriptorStatus = eService.descriptors.find(_.id == descriptorId).map(_.state)
 
-  def flattenAttributes(attributes: Seq[Attribute]): Future[Seq[AttributeValue]] = Future.fromTry {
-    Try {
-      attributes
-        .flatMap(attribute => attribute.group.toSeq.flatten ++ attribute.single.toSeq)
-    }
+    // Not using whenA on Future.failed because it requires an ExecutionContext, which is not actually needed here
+    Left[DescriptorNotInExpectedState, Unit](DescriptorNotInExpectedState(eService.id, descriptorId, allowedStates))
```

Comment:
why `Left[DescriptorNotInExpectedState, Unit](_).withRight[Unit]`? Doesn't this nest 2 eithers?

Reply:
FWIW the `Either.left[A,B]` constructor returns the left widened as a whole Either if that was the case

Reply:
maybe you are referring to `toRight`?
`withRight` just upcast the type for the right side of the `Either`. I had to add it because the compiler is not able to infer the type for the applicative for `unlessA`.
But if you have a better way in mind I'm all ears, I don't like the current solution

Reply:
never mind, your suggestion works great 🚀